File: java-rent/src/main/java/de/richargh/billiondollar/RentUseCase.java
package de.richargh.billiondollar;

import java.util.Optional;

public class RentUseCase {

    private final Inventory inventory;
    private final Renters renters;
    
    public RentUseCase(Inventory inventory, Renters renters) {
        this.inventory = inventory;
        this.renters = renters;
    }

    public boolean rent(ItemId itemId, RenterId renterId){
        boolean isRented = inventory
            .findById(itemId)
            .flatMap(item -> notRented(item))
            .flatMap(item -> rentScope(item, renters.findById(renterId)))
            .map(rentScope -> inventory.rent(rentScope.item(), rentScope.renter().id()))
            .isPresent();
            
        return isRented;
    }

    private Optional<Item> notRented(Item item){
        if(item.isRented())
            return Optional.empty();
        else
            return Optional.of(item);
    }

    private Optional<RentScope> rentScope(Item item, Optional<Renter> renter){
        if(renter.isPresent())
            return Optional.of(new RentScope(item, renter.get()));
        else
            return Optional.empty();
    }

}

File: python/src/renting/RentUseCase.py
from typing import cast

from src.renting.Inventory import Inventory
from src.renting.Renters import Renters
from src.renting.ItemId import ItemId
from src.renting.RenterId import RenterId
from src.renting.items import AvailableItem


class RentUseCase:
    _inventory: Inventory
    _renters: Renters

    def __init__(self, inventory: Inventory, renters: Renters):
        self._inventory = inventory
        self._renters = renters

    def rent(self, item_id: ItemId, renter_id: RenterId):
        item = self._inventory.find_by_id(item_id)
        renter = self._renters.find_by_id(renter_id)
        if (item is None
                or renter is None
                or not isinstance(item, AvailableItem)):
            return False
        item = cast(AvailableItem, item)

        self._inventory.rent(item, renter_id)
        return True
